### PR TITLE
Refactoring: events

### PR DESCRIPTION
**Pull request by viktorprogger**
Events refactoring.

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Tests pass?   | ❌
| Fixed issues  | #22

### CODE REVIEW

File: src/DriverInterface.php [hunk 1]
```diff
@@ -26,18 +29,25 @@ public function status(string $id): int;
...
      * @param JobInterface $job
-     * @param int $ttr time to reserve in seconds
-     * @param int $delay
-     * @param int|null $priority
      *
-     * @return string id of a job message
+     * @return MessageInterface
      */
-    public function pushMessage(JobInterface $job, int $ttr, int $delay, ?int $priority): string;
+    public function pushMessage(JobInterface $job): MessageInterface;
```
Comment (samdark) [inline src/DriverInterface.php]: What do you think about renaming pushMessage to pushJob or push?

File: src/Processors/Worker.php [hunk 2]
```diff
@@ -37,36 +33,23 @@ public function __construct(
      */
     public function process(MessageInterface $message, Queue $queue): void
     {
-        $event = ExecEvent::before($message->getId(), $message->getJob(), $message->getTtr(), $message->getAttempt());
-
-        $title = $this->formatter->getExecTitle($event);
-        $this->logger->info("$title is started.");
+        /** @var BeforeExecutionInterface $event */
+        $event = $this->factory->create(BeforeExecutionInterface::class, [$queue, $message]);
```
Comment (samdark) [inline src/Processors/Worker.php]: What's the use case to configure event classes via factory?

Comment (viktorprogger) [reply src/Processors/Worker.php]: It's a good way to make your own implementation of these events specifically for your project.

File: src/Queue.php [hunk 3]
```diff
@@ -9,15 +9,16 @@
 namespace Yiisoft\Yii\Queue;
 
 use Psr\EventDispatcher\EventDispatcherInterface;
-use Psr\Log\LoggerInterface;
-use Yiisoft\EventDispatcher\Dispatcher;
 use Yiisoft\EventDispatcher\Provider\Provider;
```
Comment (xepozz) [inline src/Queue.php]: `yiisoft/event-dispatcher` is not found in composer.json

File: src/Events/AfterExecution.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace Yiisoft\Yii\Queue\Events;
+
+use Yiisoft\Yii\Queue\MessageInterface;
+use Yiisoft\Yii\Queue\Queue;
+
+class AfterExecution implements AfterExecutionInterface
+{
+    /**
+     * @var Queue
```
Comment (xepozz) [inline src/Events/AfterExecution.php]: This and similar other PHPDocs should be removed.

File: src/Processors/Worker.php (hunk 2 above)
Comment (terabytesoftw) [reply src/Processors/Worker.php]: Wouldn't it be better to implement a factory like in widgets than to inject the entire container into the class?

Comment (viktorprogger) [reply src/Processors/Worker.php]: And what are the pros of doing so?

Comment (terabytesoftw) [reply src/Processors/Worker.php]: That I could call the factory from anywhere you need, without the need to inject it to the class.

File: src/Queue.php (hunk 3 above)
Comment (viktorprogger) [reply src/Queue.php]: It ms added in #29

**Comment by viktorprogger** [comment]
> Can u make tests for these events? I think it will be easy.

Do you mean just class unit tests or integrational tests with ordinary event creating and dispatching?

**Comment by xepozz** [comment]
As you want. 
Integrational is more important for now code base state

File: src/Processors/Worker.php (hunk 2 above)
Comment (viktorprogger) [reply src/Processors/Worker.php]: Still didn't get you :( Could you provide an example of "calling the factory from anywhere"?

Comment (samdark) [reply src/Processors/Worker.php]: > your own implementation of these events specifically for your project

Could you provide an example when you need so?

Comment (viktorprogger) [reply src/Processors/Worker.php]: - Event can provide some extended functionality, i.e. smthng like `isPropagationStopped` for the current specific case (`isSomethingHappened`).
- You can extend default event class with your own one in order to subscribe just for this implementation, so in this case the event will be cought by all the default listeners and this one without any `if` inside (viable for monolith projects with several queues):

```php
class OwnEvent extends DefaultEvent {}

$listener = fn(OwnEvent::class) => null;

$listener->subscribe($listener);
```

Comment (viktorprogger) [reply src/Processors/Worker.php]: Decided to use `new class()` instead of factory in a private discussion with @samdark.

**Comment by samdark** [comment]
Thanks!

**Pull request merged**